Introduce "InternalUser" class

tvernum opened the pull request:
This commit introduces a new `InternalUser` base class for all internal users within Elasticsearch security.

It cleans up the `User` and `InternalUsers` and `Subject` classes to interact with the new `InternalUser` class in place of the old checks based on `User.isInternal`

tvernum committed:
[WIP] Addtional refactor of Internal Users

tvernum committed:
Cleanup and fix tests

tvernum committed:
Make distinction local/remote role descriptors

tvernum committed:
Merge branch 'main' into internal-user-refactor2

tvernum committed:
Merge branch 'main' into internal-user-refactor2

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

tvernum commented:
@ywangd, no rush on this one. Please review when you're not too busy.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java:
```diff
@@ -273,12 +274,16 @@ private RoleReferenceIntersection buildRoleReferencesForCrossClusterAccess() {
+                        "The internal user [" + internalUser + "] is not permitted to perform cross cluster actions"
+                    )
+                );
+            roleReferences.add(new RoleReference.FixedRoleReference(internalRoleDescriptor, "cross_cluster_access_internal"));
```
@ywangd Please check that what I've done here looks right to you.

ywangd replied:
Yeah this looks right.

ywangd reviewed (commented):
I only looked at changes related to `InternalUser` briefly. But I have a question around `InternalUsers` which could use early discussions. Thanks!

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/InternalUsers.java:
```diff
@@ -7,94 +7,73 @@
-import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
 public class InternalUsers {
```
I wondered whether you'd say this (though it didn't really occur to me until I looked at the diff in GitHub).

If you're happy with a bigger PR, I'm happy to push the refactor further.

ywangd replied:
Yep I am OK with a bigger change along this line. I don't think reservedUser has to be part of it. But that does not seem to add much anyway. Thanks!

tvernum replied:
Yeah, I really only added `ReservedUser` so that all in-code users have an explicit type hierarchy. They're either `ReservedUser` or `InternalUser`, never just a `User`.
It makes it easier to look at a class and know what type of user we're dealing with.

tvernum committed:
Fix typos in javadoc

Co-authored-by: Yang Wang <ywangd@gmail.com>

tvernum committed:
Merge branch 'main' into internal-user-refactor2

tvernum committed:
Remove 'isInternal' and 'getInternalUserName' methods

tvernum committed:
Remove role related methods

tvernum committed:
Remove independent classes for internal users

This commit drops (most) classes that extends InternalUser and
replaces them with simple instances fields in the InternalUsers class.

tvernum committed:
Merge branch 'main' into internal-user-refactor2

tvernum committed:
Cleanup

tvernum committed:
Merge branch 'main' into internal-user-refactor2

tvernum committed:
Merge branch 'main' into internal-user-refactor2

tvernum commented:
@ywangd I've made all the changes that I think are necessary.

There are still classes for `InternalUsers`, `SystemUser` and `CrossClusterAccessUser`
It's possible to remove some/all of those, but I'm not convinced we end up with a better codebase from doing so.

I'm happy to do more work if you think there's value, but I feel like this has landed in the right place.

ywangd commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/InternalUserAndRoleIntegTests.java:
```diff
@@ -13,34 +13,18 @@
-        UsernamesField.CROSS_CLUSTER_ACCESS_ROLE };
+
+    private static final String[] INTERNAL_USERNAMES = InternalUsers.get().stream().map(User::principal).toArray(String[]::new);
+    private static final String[] INTERNAL_ROLE_NAMES = INTERNAL_USERNAMES;
```
Seeing this makes me wonder whether we should have an assertion in InternalUser constructor to ensure this is true?

ywangd commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizationServiceTests.java:
```diff
@@ -2582,9 +2582,10 @@ private Authentication createAuthentication(User user) {
+        if (user instanceof InternalUser internalUser) {
             assert authenticatingUser == null;
-            authentication = AuthenticationTestHelper.builder().internal(user).build();
+            assert internalUser instanceof InternalUser;
```
This is redundant since `internalUser` is already casted.

ywangd reviewed (commented):
This is looking great. I don't expect to come up with any major comments. But would like to have another pass through the changes. Meanwhile, here are the some comments and questions.

Also, do you still have the view that CrossClusterAccessUser should be merged into SystemUser? The local/remote roleDescriptor structure lends itself very well to this change. If so, I can raise a follow-up PR for it.

tvernum committed:
Merge branch 'main' into internal-user-refactor2

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/ReservedUser.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.util.Optional;
+
+public abstract class ReservedUser extends User {
```
I think so. I don't think there's any reason to have a new class for each reserved user.
But I definitely don't want to complicate this PR by doing that.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/CrossClusterAccessUser.java:
```diff
@@ -39,24 +39,18 @@ public class CrossClusterAccessUser extends User {
+        );
     }
 
     public static boolean is(User user) {
```
I think I can actually get rid of it now. It only has 1 remaining production use

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/InternalUsers.java:
```diff
@@ -7,94 +7,144 @@
-            CrossClusterAccessUser::is
-        );
-        defineUser(StorageInternalUser.NAME, StorageInternalUser.INSTANCE, StorageInternalUser.ROLE_DESCRIPTOR, StorageInternalUser::is);
+        defineUser(SystemUser.INSTANCE);
```
I don't have a strong view either way. I can see an argument for all internal users being in 1 place, but it means the `SystemUser` constructor needs to become package protected instead of private.

ywangd replied:
A middle ground is probably just create a copy for `SystemUser.INSTANCE` in `InternalUsers`, i.e. `InteralUsers.SYSTEM_USER = SystemUser.INSTANCE`. What do you think?

tvernum committed:
Address feedback

tvernum commented:
@ywangd I've addressed your earlier feedback, did you want to do another round?

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/InternalUser.java:
```diff
@@ -0,0 +1,65 @@
+     * originate from a node within an external cluster (via CCS/CCR).
+     * @see #getLocalClusterRoleDescriptor()
+     */
+    public Optional<RoleDescriptor> getRemoteAccessRole() {
```
Sorry to drag on this. But could we rename this to `getRemoteAccessRoleDescriptor`?

tvernum replied:
Sorry, I thought I did both, some distraction must have happened midway.

ywangd reviewed (approved):
LGTM

tvernum committed:
Address additional feedback

tvernum committed:
Fix * imports

tvernum committed:
Merge branch 'main' into internal-user-refactor2
